File: javascript/lib/filelock.js
// cross-process advisory lock
//
// the promise-chain mutex in store.js serializes work inside one node
// process. two separate invocations, which is what two users running the cli
// at the same time looks like, share no such state and will interleave their
// read-modify-write cycles freely.
//
// the lock is a file created with the exclusive flag, which fails if the path
// already exists. that check and create is atomic at the syscall level, so
// exactly one process wins. java acquires the same lock by the same means,
// which is why the mechanism is fixed in the spec: an advisory lock only
// works when every participant honors it the same way
const fs = require('fs/promises');
const path = require('path');

const LOCK_FILE = path.join(__dirname, '..', '..', 'data', 'tasks.lock');
const RETRY_MS = 5;
const TIMEOUT_MS = 10000;
const STALE_MS = 30000;

class LockError extends Error {
  constructor(message) {
    super(message);
    this.code = 3;
  }
}

const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

// breakIfStale removes a lock left behind by a process that died holding it
async function breakIfStale(file) {
  try {
    const stat = await fs.stat(file);
    if (Date.now() - stat.mtimeMs > STALE_MS) {
      await fs.unlink(file);
      return true;
    }
  } catch {
    // the holder released it while we were looking, which is fine
  }
  return false;
}

async function acquire(file = LOCK_FILE) {
  const deadline = Date.now() + TIMEOUT_MS;

  for (;;) {
    try {
      // wx fails if the path exists, so the winner is decided by the kernel
      const handle = await fs.open(file, 'wx');
      await handle.writeFile(String(process.pid));
      await handle.close();
      return;
    } catch (err) {
      if (err.code !== 'EEXIST') {
        throw new LockError(`cannot acquire ${file}: ${err.message}`);
      }
      if (Date.now() > deadline) {
        if (await breakIfStale(file)) continue;
        throw new LockError(`timed out waiting for ${file} after ${TIMEOUT_MS}ms`);
      }
      await sleep(RETRY_MS);
    }
  }
}

async function release(file = LOCK_FILE) {
  try {
    await fs.unlink(file);
  } catch (err) {
    if (err.code !== 'ENOENT') {
      throw new LockError(`cannot release ${file}: ${err.message}`);
    }
  }
}

// withLock guarantees release even when fn throws
async function withLock(fn, file = LOCK_FILE) {
  await acquire(file);
  try {
    return await fn();
  } finally {
    await release(file);
  }
}

module.exports = { LOCK_FILE, LockError, acquire, release, withLock };

File: java/src/todo/store/FileLock.java
package todo.store;

import java.io.IOException;
import java.nio.file.FileAlreadyExistsException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.attribute.FileTime;

/**
 * Cross process advisory lock.
 *
 * The lock is a file created with CREATE_NEW, which fails if the path already
 * exists. That check and create is atomic at the syscall level, so exactly one
 * process wins. The JavaScript implementation acquires the same lock the same
 * way. FileChannel.lock would be the more idiomatic Java choice, but an OS
 * advisory lock and an exclusive create do not exclude one another, so the
 * mechanism is fixed in the spec rather than chosen per language.
 */
public final class FileLock implements AutoCloseable {

    private static final long RETRY_MS = 5;
    private static final long TIMEOUT_MS = 10_000;
    private static final long STALE_MS = 30_000;

    private final Path path;

    private FileLock(Path path) {
        this.path = path;
    }

    /** Blocks until the lock is held, then returns a handle that releases it. */
    public static FileLock acquire(Path path) throws IOException {
        long deadline = System.currentTimeMillis() + TIMEOUT_MS;

        while (true) {
            try {
                Files.createFile(path);
                Files.writeString(path, String.valueOf(ProcessHandle.current().pid()));
                return new FileLock(path);
            } catch (FileAlreadyExistsException e) {
                if (System.currentTimeMillis() > deadline) {
                    if (breakIfStale(path)) {
                        continue;
                    }
                    throw new IOException("timed out waiting for " + path);
                }
                try {
                    Thread.sleep(RETRY_MS);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    throw new IOException("interrupted waiting for " + path, ie);
                }
            }
        }
    }

    // removes a lock left behind by a process that died holding it
    private static boolean breakIfStale(Path path) {
        try {
            FileTime modified = Files.getLastModifiedTime(path);
            if (System.currentTimeMillis() - modified.toMillis() > STALE_MS) {
                Files.deleteIfExists(path);
                return true;
            }
        } catch (IOException ignored) {
            // the holder released it while we were looking, which is fine
        }
        return false;
    }

    /** try with resources guarantees release even when the body throws */
    @Override
    public void close() throws IOException {
        Files.deleteIfExists(path);
    }
}
